### DISCUSSION TITLE
Opt out of stale-while-revalidate behavior

### DISCUSSION THREAD

**Discussion by assafweinberg** [vercel/next.js#56050] [Ideas]
### Goals

1. Opt out of the default `stale-while-revalidate` data caching behavior
2. Choose caching behavior on fetch request
3. Allow data cache to guarantee non-stale data is returned.

### Background

Sometimes it's better for stale cache hits to behave as a MISS rather than following the `stale-while-revalidate` process. This happens in situations where stale data is known to be invalid.

For example, some 3rd party APIs return fields that expire like image URLs. See Airtable's API behavior for image urls for example - https://community.airtable.com/t5/development-apis/breaking-change-urls-to-attachments-have-been-modified/td-p/79821

NextJS's default `stale-while-invalidate` isn't ideal in these situations because it leads to some users seeing broken images or getting invalid paging errors on infrequently visited pages. Without being able to change the `stale-while-revalidate` behavior, the only other option is to turn off the cache entirely. There should be an option to use caching but guarantee that the data returned is not stale. 

Ideally, you should be able to specify caching behavior by request when calling fetch or at least get an indication in the response that the data being returned is stale.

### Proposal

Something like...

```
fetch(url, {
    next: { 
        revalidate: 60, 
        cacheBehavior: 'revalidate-if-stale' | 'stale-while-revalidate'
    },
  });
```

**Comment by felipebarichello** [comment]
EDIT: See my own reply to this comment for a workaround I found.

---

I feel the need for that feature should have been obvious from the beggining. After discovering Next, this was one of the first things I searched about: how Next caching works and the options I have to configure it. Agressive caching is the main appeal for the use of Next, at least for me. In fact the ISR is the only reason I'm going for Next. And in that search, finding that I can't avoid having two month old critical data was mind-blowing.

Maybe there is a technical reason that prevents that feature from existing without performance hits. I'd like to know if that's the case.

Consider the following example (I've met a situation similar to that, but I replaced the context by sweepstakes for privacy and simplicity):

Let's say it is a website of sweepstake results. And for many sweepstakes pages (promotional sweepstakes from companies A and B), the results are accessed hundreds of times every minute. But some sweepstake pages (from company C) are really not popular on this website and users frequently access those pages only the next day to see if they won. If their sweepstake result is not there by the next day, they will probably never trust that website again. But also I can't afford SSR because of the amount of requests on the popular sweepstake pages. All company pages use the same page component. CSR is also not an option because people searching for the date of their sweepstake result should be able to find that page (SEO is very critical for that website).

Is there a workaround to that problem at least? And I mean besides the obvious external tool fetching pages acting as users, because that may create unwanted load in the server if many pages are rarely accessed for instance.

**Reply by felipebarichello** [reply]
It seems that <https://nextjs.org/docs/app/api-reference/functions/cacheLife#stale> has the workaround feature. You can cache at the function/component level, and with `cacheLife` you can have a pretty decent cache control over when to cache client-side, when to stale-while-revalidate and when to revalidate-if-stale, all at the same time. For your fetches, you can wrap your fetch within a function and disable the fetch cache. It's more verbose, but should work.
